### ISSUE TITLE
Pickling of pipelines

### ISSUE THREAD

**Issue by MagBuchSB1** [pdpipe/pdpipe#71]
Hi,

I stumbled upon your excellent library when researching ways of serializing and pickling pipelines for reuse on new datasets, and found your approach to be most promising. 
However, when attempting to pickle to the following pipeline from your documentation;

`pipeline = pdp.ColDrop("Name") + pdp.OneHotEncode("Label")`
`pipeline += pdp.MapColVals("Job", {"Part": True, "Full":True, "No": False})`
`pipeline += pdp.PdPipeline([pdp.ColRename({"Job": "Employed"})])`
`joblib.dump(pipeline, "pipeline.pkl")`

... I get the following Exception:

`PicklingError: Can't pickle <function ColRename.__init__.<locals>._tprec at 0x000001CBC33864C0>: it's not found as pdpipe.basic_stages.ColRename.__init__.<locals>._tprec`

I get similar exceptions when attemtping other function-based steps, like ApplyToRows or ColByFrameFunc.

Is there a way to get around this? As an example: How would I solve applying a function that calculates age from a column containing birthdate and pickling this?

In general I'd love to see more explanations and examples on pickling pipelines in your otherwise excellent documentation!

**Comment by MagBuchSB1** [comment]
Apologies, seems like https://github.com/pdpipe/pdpipe/issues/45 addresses the same issue.

**Comment by shaypal5** [comment]
No no!

This is definitely a bug on my end! I'll look into it the coming days.
All these pipeline stages should pickle out of the box, and they did in previous versions.

I think that perhaps some new feature that has something todo with that `_tprec` private attribute broke this.

Which also means I have some missing tests!

I'll take a look in the coming days and get back to you. This should be an easy fix. I serialized pipelines before for actual code projects in production for my clients. We can do this. 😎

**Comment by MagBuchSB1** [comment]
Great @shaypal5, looking forward to hearing from you!

**Comment by MagBuchSB1** [comment]
As for my other question on calculating age. Here's one way of solving this with a lambda:
```age_test = pd.DataFrame({'BORN_DATE': ['1/1/1980', '1/31/1972']})
age_test['BORN_DATE'] = pd.to_datetime(age_test['BORN_DATE'])
age_func = lambda row: date.today().year - row['BORN_DATE'].year - ((date.today().month, date.today().day) < (row['BORN_DATE'].month, row['BORN_DATE'].day))
add_age = pdp.ApplyToRows(age_func, 'AGE')
add_age(age_test)
```
![image](https://user-images.githubusercontent.com/36260162/143877521-8c6553ce-2765-40bd-9bd6-33ba5cbfe3b1.png)

This works fine, but cannot be pickled:
`joblib.dump(add_age, "add_age.pkl")` yields:
`PicklingError: Can't pickle <function <lambda> at 0x000001CBC4177D30>: it's not found as __main__.<lambda>`

Is there an alternative "picklable" way of achieving this?

**Comment by shaypal5** [comment]
Yes. Named function, even private ones (meaning, ones defined inside scopes, like inside functions/methods or inside `for` loops) are pickle-able.

This should work:
```python
age_test['BORN_DATE'] = pd.to_datetime(age_test['BORN_DATE'])

def age_func(row):
  return date.today().year - row['BORN_DATE'].year - ((date.today().month, date.today().day) < (row['BORN_DATE'].month, row['BORN_DATE'].day))

add_age = pdp.ApplyToRows(age_func, 'AGE')
add_age(age_test)
```

**Comment by shaypal5** [comment]
Also, writing the name of a programming language after the three ``` that start a code block makes it syntax-highlighted.

Meaning:
```
```python
  def foo():
    return 5
\```
```
(without the `\`)

will render as:
```python
  def foo():
    return 5
```

And `bash` is also supported.

**Comment by MagBuchSB1** [comment]
Thanks for the syntax highlighting tip, I knew there was a better way to post code. I'll be better :)

As for your example, I cannot seem to get that to work unless I pickle and unpickle the function separately.  

First dump:

```python
import dill

def age_func(row):
    return date.today().year - row['BORN_DATE'].year - ((date.today().month, date.today().day) < (row['BORN_DATE'].month, row['BORN_DATE'].day))
# Dump the function    
with open('age_func.pkl', 'wb') as file:
    dill.dump(age_func, file) 

add_age = pdp.ApplyToRows(age_func, 'AGE')
# dump pdp
joblib.dump(add_age, "add_age.pkl")
```

Then load and apply:

```python
age_test = pd.DataFrame({'BORN_DATE': ['1/1/1980', '1/31/1972']})
age_test['BORN_DATE'] = pd.to_datetime(age_test['BORN_DATE'])

with open('age_func.pkl', 'rb') as f:
    age_func = dill.load(f)

add_age = joblib.load("add_age.pkl")
result = add_age(age_test)
```

I'm probably messing something up. This is, however, close to an acceptable workaround, although less dynamic than I'd wished.

**Comment by MagBuchSB1** [comment]
Here's a thought: 
In order to make sure the named function is available after serializing and deserializing the pipe, how about serializing and deserializing the function within the pdpipe class? This seems to work in my tests. Here's how I tested it:

In class ApplyByRows.\__init__, included:

```python
self._named_func = pickle.dumps(func) # using import dill as pickle
```
Then in ApplyByRows._transform:
```python
named_func = pickle.loads(self._named_func)
new_cols = df.apply(named_func, axis=1)
```
Obviously not robust like this, but maybe an approach to consider?

**Comment by MagBuchSB1** [comment]
On further investigation it seems like cloudpickle might solve this, as it serializes by value and not by only reference.

**Comment by shaypal5** [comment]
Solved!

Changed the inner implementation to use `Condition` objects instead of local functions. Way prettier, too. I'll release a version in the next 24 hours, and then move to write tests covering serialization of all pipeline stages.

**Comment by shaypal5** [comment]
Released in `v0.0.68`. Let me know if serialization now works, or if another stage fails to serialize.

**Comment by shaypal5** [comment]
A couple of additional unpickle-able stage were found and fixed.
Released in `v0.0.69`.
Everything should be pickle-able now.
